eventsources/sources/redisStream/start.go

```diff
@@ -0,0 +1,204 @@
+	}
+	log.Infof("connected to redis server %s", redisEventSource.HostAddress)
+
+	// Create a common consumer group on all streams to start reading from beginning of the streams.
+	// Only proceeds if all the streams are already present
+	consumersGroup := "argo-events-cg"
+	if len(redisEventSource.ConsumerGroup) != 0 {
+		consumersGroup = redisEventSource.ConsumerGroup
+	}
+	for _, stream := range redisEventSource.Streams {
+		// create a consumer group to start reading from the current last entry in the stream (https://redis.io/commands/xgroup-create)
+		if err := client.XGroupCreate(ctx, stream, consumersGroup, "$").Err(); err != nil {
+			// redis package doesn't seem to expose concrete error types
+			if err.Error() != "BUSYGROUP Consumer Group name already exists" {
+				return errors.Wrapf(err, "creating consumer group %s for stream %s on host %s for event source %s", consumersGroup, stream, redisEventSource.HostAddress, el.GetEventName())
+			}
+			log.Infof("Consumer group %s already exists", stream)
+		}
+	}
+
+	readGroupArgs := make([]string, 2*len(redisEventSource.Streams))
+	copy(readGroupArgs, redisEventSource.Streams)
+	// Start by reading our pending messages(previously read but not acknowledged).
+	streamToLastEntryMapping := make(map[string]string, len(redisEventSource.Streams))
+	for _, s := range redisEventSource.Streams {
+		streamToLastEntryMapping[s] = "0-0"
+	}
+
+	updateReadGroupArgs := func() {
+		for i, s := range redisEventSource.Streams {
+			readGroupArgs[i+len(redisEventSource.Streams)] = streamToLastEntryMapping[s]
+		}
+	}
+	updateReadGroupArgs()
+
+	msgCount := redisEventSource.MaxMsgCountPerRead
+	if msgCount == 0 {
+		msgCount = 10
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			log.Infof("Redis stream event source for host %s is stopped", redisEventSource.HostAddress)
+			return nil
+		default:
+		}
+		entries, err := client.XReadGroup(ctx, &redis.XReadGroupArgs{
+			Group:    consumersGroup,
+			Consumer: "argo-events-worker",
+			Streams:  readGroupArgs,
+			Count:    int64(msgCount),
+			Block:    2 * time.Second,
+			NoAck:    false,
+		}).Result()
+		if err != nil {
+			if err == redis.Nil {
+				continue
+			}
+			return errors.Wrapf(err, "reading streams %s using XREADGROUP", strings.Join(redisEventSource.Streams, ", "))
+		}
+
+		for _, entry := range entries {
+			if len(entry.Messages) == 0 {
+				// Completed consuming pending messages. Now start consuming new messages
+				streamToLastEntryMapping[entry.Stream] = ">"
+			}
+			for _, message := range entry.Messages {
+				if err := el.handleOne(entry.Stream, message, dispatch, log); err != nil {
+					log.With("stream", entry.Stream, "message_id", message.ID).Errorw("failed to process Redis stream message", zap.Error(err))
+					el.Metrics.EventProcessingFailed(el.GetEventSourceName(), el.GetEventName())
+					continue
+				}
+				if err := client.XAck(ctx, entry.Stream, consumersGroup, message.ID).Err(); err != nil {
```

Comment:
Since doing a batch read, this could be enhanced to `XAck` a list of message.ID.